Add lazy dropEnd and friends

3kyro opened the pull request:
Fixes #306.
Utilizes an eager strategy as described in the original issue. Chunks
are eagerly evaluated but the overall structure and pointers are kept
intact.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -684,6 +690,30 @@ take i cs0         = take' i cs0
+-- >>> takeEnd 4 "abc"
+-- "abc"
+--
+-- @since 0.11.1.1
```
```suggestion
-- @since 0.11.2.0
```
Here and below please.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -733,6 +804,23 @@ dropWhile f = dropWhile'
+            fromStrict (S.dropWhileEnd f c)
+        dropWhileEnd' (Chunk c bs) =
+            case dropWhileEnd' bs of
+                 Empty -> dropWhileEnd' (Chunk c Empty)
```
This is a bit entangled, to be honest. If you put `fromStrict (S.dropWhileEnd f c)` instead of `dropWhileEnd' (Chunk c Empty)`, you'll be able to remove special handling of `Chunk c Empty` case.

3kyro commented:
Thanks for the help!

I used `foldrChunks` instead of `length` based checks, but I still have some concerns/questions: Will the tuple based folds work when used in a streaming manner, as discussed in a [comment](https://github.com/haskell/bytestring/issues/306#issuecomment-711395956) in the original issue? e.g. would ``take x . dropEnd y`` operate in constant space?

Would explicitly implementing the folds help? eg;
```` haskell
dropEnd i cs0 = cs 
  where 
    (_,cs) = go i cs0
    go :: Int64 -> ByteString -> (Int64, ByteString)
    -- n keeps track of dropped bytes    
    go n cs = ...
````

Bodigrim commented:
Sorry, @3kyro, I'm on the go and won't have time to review until two weeks from now.

> Will the tuple based folds work when used in a streaming manner, as discussed in a [comment](https://github.com/haskell/bytestring/issues/306#issuecomment-711395956) in the original issue?

Good question. Could you possibly check it in REPL with something like `Chunk "foo" (Chunk "bar" undefined)`? Bonus points for adding tests, you can find some inspiration [here](https://github.com/haskell/bytestring/pull/364/files#diff-cdcae7a31dfaa8744331575bdfbfeaaed28d42fbe558e79b8fd7312a5061fde0R469).

3kyro commented:
No problem. I'll check lazy behavior and add some checks as well as fixing ghc-8,2 CI failures.

3kyro commented:
Hi @Bodigrim, I'm not really sure I can create a version of these end manipulation functions that is sufficiently lazy. Based on the discussion in #306, we evaluate all chunks when doing `(take x . dropEnd y) cs`. This seems necessary as `cs` needs to be fully evaluated to know if there's anything left to `take`.

Regarding style, I'm not sure if you'd prefer a more explicit implementation, similar to other functions in the module (`drop`, `take` etc). `dropEnd` would be something like: 
````haskell
dropEnd :: Int64 -> ByteString -> ByteString
dropEnd i p | i <= 0 = p
dropEnd i cs0 = cs0' 
  where (_ , cs0') = dropEnd' (fromIntegral i) cs0
        dropEnd' n cs | n <= 0 = (0, cs)
        dropEnd' n Empty = (n, Empty)
        dropEnd' n (Chunk c Empty) = 
            (n - S.length c, S.fromStrict $ S.dropEnd n c)
        dropEnd' n (Chunk c cs) = 
            let (n',cs') = dropEnd' n cs
              in (n' - S.length c, S.fromStrict (S.dropEnd n' c) `append` cs')
````

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -684,6 +690,28 @@ take i cs0         = take' i cs0
+takeEnd i cs0        = takeEnd' i cs0
+  where takeEnd' 0 _         = Empty
+        takeEnd' n cs        =
+            snd $ foldrChunks takeTuple (fromIntegral n,Empty) cs
```
On 32-bit platforms this `fromIntegral` converts `Int64` to `Int32`. Please do it other way around: pass `Int64` in accumulator, but subtract `fromIntegral (S.length c)`.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -750,6 +832,23 @@ break f = break'
+          case breakEnd' cs of
+               (Empty, cs') ->
+                 let (c', c'') = S.breakEnd f c
+                   in (fromStrict c', fromStrict c'' `append` cs')
```
```suggestion
                   in (fromStrict c', c'' `chunk` cs')
```
Maybe even ``(c' `chunk` Empty, c'' `chunk` cs'`)`` for more symmetry.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -733,6 +800,21 @@ dropWhile f = dropWhile'
+-- @since 0.11.2.0
+dropWhileEnd :: (Word8 -> Bool) -> ByteString -> ByteString
+dropWhileEnd f = dropWhileEnd'
+  where dropWhileEnd' Empty        = Empty
```
I suggest the following strategy. Start with an empty accumulator. If you see `Chunk c bs`, check whether the last char of `c` satisfies `f`. If it does not, you can immediately return accumulated strings and `c` - without inspecting `bs`! If the last char of `c` satisfies `f`, append `c` to the accumulator and proceed with `bs`. Further, when you finally encounter `Empty`, remove from accumulator a tail satisfying `f` (which may span over several chunks) and return it.

3kyro replied:
> 
> 
> Sorry, @3kyro, I got distracted by other things for longer than planned. Hope it was not too long.

Hi @Bodigrim , there's absolutely no problem. I'm trying to wrap my ahead around your suggestion, but I'm not up to the task. Considering a function
````haskell
go (Chunk c cs) acc
    | f (S.last c) = go cs (acc `append` fromStrict c)
    | otherwise = acc `append` fromStrict c
```` 
Assuming we start from the left, I fail to see how I can not inspect `cs` if `f (S.last c)`  is `False`. How do we know that there isn't a `Word8` in `cs` that does not satisfy `f`? 

You also mention
> when you finally encounter Empty, remove from accumulator a tail satisfying f (which may span over several chunks) and return it.

doesn't this need an implenatation of `dropWhileEnd` that we are in the process of writing?

I'm certainly missing something obvious here so I would really appreciate your help.

Bodigrim replied:
Sorry, I was not talking clear. I meant something like this:
```haskell
go :: [S.ByteString] -> ByteString -> ByteString
go acc (Chunk c cs) 
  | f (S.last c) = go (c : acc) cs 
  | otherwise    = foldl (flip Chunk) (c : acc) (go [] cs)
```
In the second case we can start constructing lazy `ByteString` before forcing `go [] cs`. This is important, because if a consumer cares only about the beginning of the string (e. g., `take n . dropWhileEnd f`), `go [] cs` could remain unevaluated.

You are absolutely right, for `go acc Empty` we basically employ an eager algorithm you have written for `dropWhile` above (note that chunks in `acc` are in reversed order, however).

3kyro replied:
Makes absolute sense, thanks a lot.

I've added your suggestions plus a lazy test for `dropWhileEnd`

3kyro replied:
`breakEnd` and `spanEnd` updated

Bodigrim reviewed (commented):
Sorry, @3kyro, I got distracted by other things for longer than planned. Hope it was not too long.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +725,28 @@ drop i cs0 = drop' i cs0
+-- @since 0.11.2.0
+dropEnd :: Int64 -> ByteString -> ByteString
+dropEnd i p | i <= 0 = p
+dropEnd i cs0 = dropEnd' i cs0
```
I suggest we make `dropEnd` lazier in a fashion vaguely similar to `dropWhileEnd`. Imagine `dropEnd 10` and chunks of size 1, 2, 3, 6... As soon as you encountered the fourth chunk, you know that you can output the first one no matter what. Because 1 <= 2 + 3 + 6 - 10.

3kyro replied:
Hi @Bodigrim , `dropEnd` updated. I'm slowly  getting there 😄

Bodigrim commented:
@3kyro I'm afraid this is not quite correct, as you can see from CI. Try 
```
cabal test --test-options '-p dropEnd --quickcheck-tests 1000'
```

3kyro commented:
Hi @Bodigrim, sorry for the delay but I was away for some days.

I have corrected the behavior of `dropEnd`, trying not to reverse chunk lists too much. I believe I'm still within O(n) but I would really appreciate your feedback. Having a FIFO container would make the code much simpler, but I guess importing `containers` is out of the question.

sjakobi commented:
> Having a FIFO container would make the code much simpler, but I guess importing `containers` is out of the question.

AFAIK adding a dependency on `containers` (or maybe alternatively `array`) would probably be okay. We'd have to check how it affects GHC build times though.

Bodigrim commented:
I think this PR does not justify an additional dependency, even `containers`. We do not really need `Data.Sequence` here, only a simple stack, which can be implemented in a small utility module (as two lists). But I guess reversing is not prohibitively expensive.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,43 @@ drop i cs0 = drop' i cs0
+        getOutput out [] [] acc = (out, [], [], acc)
+        getOutput out [] bss acc = getOutput out (L.reverse bss) [] acc
+        getOutput out (x:xs) bss acc =
+            if len x <= acc - i - len x
```
Looks suspicious, there are `len x` on both sides. Wouldn't `len x <= acc - i` be enough?

3kyro replied:
Yes, its redundant

Bodigrim reviewed (commented):
@3kyro, nice! I think we are almost there, please take a look at suggestions.

3kyro commented:
Hi @Bodigrim, I've added an unexposed `Data.Bytestring.Lazy.Internal.Deque` module and modified `dropEnd` accordingly. I have not prepared tests for `Deque`'s functions as the module only tests the exposed API.

3kyro commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,43 @@ drop i cs0 = drop' i cs0
+        len c = fromIntegral (S.length c)
+        hLen cs = maybe 0 len (listToMaybe cs)
+
+        getOutput out [] [] acc = (out, [], [], acc)
```
I believe now with the use of the Deque, arguments are quite easy to follow. I'd happily add comments though if necessary

3kyro commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,43 @@ drop i cs0 = drop' i cs0
+dropEnd :: Int64 -> ByteString -> ByteString
+dropEnd i p | i <= 0 = p
+dropEnd i p = go [] [] 0 0 p
+  where go hss tss acc h (Chunk c cs)
```
I have removed the need for the `head` of `hss` as its length is already included  in the accumulated length. Furthermore, in a Deque, calculating the length of its`head` might trigger a re-balancing which I think would not be ideal (e.g. it would always re balance on the second chunk). This seems a bit cleaner, although we might trigger `getOutput` a bit more often than before.

Bodigrim commented on Data/ByteString/Lazy/Internal/Deque.hs:
```diff
@@ -0,0 +1,60 @@
+-- O(1) , occasionally O(n)
+popFront :: Deque -> (Maybe S.ByteString, Deque)
+popFront (Deque [] [] _) = (Nothing, empty)
+popFront (Deque [] rs acc) = popFront (Deque (reverse rs) [] acc)
```
Such pattern is appealing, but GHC is likely to conclude that there is an unbounded general recursion going on and will resist to inline and optimize and such. I'd rewrite it as 
```haskell
popFront (Deque [] rs acc) = case reverse rs of 
  [] -> ... 
  x : xs -> ...
```

Bodigrim commented on Data/ByteString/Lazy/Internal/Deque.hs:
```diff
@@ -0,0 +1,60 @@
+-- Pop a `S.ByteString` from the front of the `Deque`
+-- Returns the bytestring, or Nothing if the Deque is empty, and the updated Deque
+-- O(1) , occasionally O(n)
+popFront :: Deque -> (Maybe S.ByteString, Deque)
```
```suggestion
popFront :: Deque -> Maybe (S.ByteString, Deque)
```
This is a bit more precise: if it's `Nothing`, there is little point to return a "new" `Deque`.

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,55 @@ drop i cs0 = drop' i cs0
+                       else (out, deque)
+
+        -- drop n elements from the rear of the accumulating `deque`
+        dropElements deque n
```
Could you please add type signatures for `go`, `getOutput`, `dropElements` and `fromDeque`?

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,55 @@ drop i cs0 = drop' i cs0
+            case D.popFront deque of
+                 (Nothing, deque') -> (out, deque')
+                 (Just x, deque')  ->
+                    if len x <= D.elemLength deque' - i
```
`deque'` is `deque` without `x`, so I'd expect just `if D.elemLength deque' >= i then ...`

Bodigrim commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,55 @@ drop i cs0 = drop' i cs0
+
+        -- drop n elements from the rear of the accumulating `deque`
+        dropElements deque n
+            | D.null deque = Empty
```
I think you do not need to handle this case specially: if `deque` is empty, then `D.popRear` returns `Nothing` and `Empty` naturally arises below.

Bodigrim reviewed (commented):
Almost there!

3kyro commented:
Latest review changes added

sjakobi commented on bytestring.cabal:
```diff
@@ -90,6 +90,7 @@ library
                      Data.ByteString.Builder.Internal
                      Data.ByteString.Builder.Prim.Internal
   other-modules:     Data.ByteString.Builder.ASCII
+                     Data.ByteString.Lazy.Internal.Deque
```
Please sort the `other-modules` in lexicographic order.

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,52 @@ drop i cs0 = drop' i cs0
+                             | otherwise  -> (out, deque)
+
+        -- drop n elements from the rear of the accumulating `deque`
+        dropElements :: D.Deque -> Int -> ByteString
```
Why is it safe to use an `Int` here, even on 32-bit systems?

```suggestion
        dropElements :: D.Deque -> Int64 -> ByteString
```

3kyro replied:
It is not safe. Corrected

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,52 @@ drop i cs0 = drop' i cs0
+            Just (x, deque') | D.elemLength deque' >= i -> getOutput (x:out) deque'
+                             | otherwise  -> (out, deque)
+
+        -- drop n elements from the rear of the accumulating `deque`
```
The term "elements" is slightly confusing in this context: Does it mean the bytes or the individual strict bytestring chunks contained in the deque? I'd suggest using the term "bytes".

3kyro replied:
I used elements as its consistently used to describe bytes in the module (eg. documentation for `drop`, `takeWhile` etc). The same applies for strict bytestirngs (e.g. S.dropEnd  also uses 'elements'). Is this reasoning sufficient?

sjakobi replied:
In this case you seem to refer to the elements of the deque. However the operations implemented for the `Deque` type all seem to handle strict bytestrings as elements. So I think that "bytes" would be clearer. Maybe you can find a different wording that clarifies things though.

sjakobi commented on Data/ByteString/Lazy/Internal/Deque.hs:
```diff
@@ -0,0 +1,62 @@
+module Data.ByteString.Lazy.Internal.Deque (
```
Please add a module header that refers to the Deque usage in `dropEnd`.

sjakobi commented on Data/ByteString/Lazy/Internal/Deque.hs:
```diff
@@ -0,0 +1,62 @@
+    { front :: [S.ByteString]
+    , rear :: [S.ByteString]
+    , -- | Accumulated length of deque's elements
+      elemLength :: Int64
```
Is this field intentionally lazy? Why?

```suggestion
      elemLength :: !Int64
```

3kyro replied:
It's not intentional. Corrected.

sjakobi commented on Data/ByteString/Lazy/Internal/Deque.hs:
```diff
@@ -0,0 +1,62 @@
+-- Pop a `S.ByteString` from the rear of the `Deque`
+-- Returns the bytestring and the updated Deque, or Nothing if the Deque is empty
+-- O(1) , occasionally O(n)
+popRear :: Deque -> Maybe (S.ByteString, Deque)
```
This type fits slightly better with my intuition, but I don't mind too much.

```suggestion
popRear :: Deque -> Maybe (Deque, S.ByteString)
```

3kyro replied:
Yes, looks nicer. thanks

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,52 @@ drop i cs0 = drop' i cs0
+
+        -- get all `S.ByteString` from the front of the accumulating deque
+        -- for which we know they won't be dropped
+        getOutput :: [S.ByteString] -> D.Deque -> ([S.ByteString], D.Deque)
```
Instead of operating on lists of strict bytestrings, it seems convenient to simply use the lazy bytestring type. Is there a reason not to do that here?

3kyro replied:
yes, seems nicer. The only problem is that there is no lazy left fold for chunks, and so we need to first reverse chunks and fold from the right. Might be even better to pay traveral than have the space leaked by the foldl?

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,52 @@ drop i cs0 = drop' i cs0
+            | D.elemLength deque < i = go (D.snoc c deque) cs
+            | otherwise              =
+                  let (output, deque') = getOutput [] (D.snoc c deque)
+                    in L.foldl (flip chunk) (go deque' cs) output
```
What's the reason for using a lazy `foldl` here instead of `foldl'`?

3kyro replied:
The lazy fold here is necessary so that we don't force `go deque` cs`. However this has now been changed (see above review comment)

sjakobi reviewed (commented):
Good progress. I'll probably be AFK until Saturday. I can review more then.

3kyro commented:
> Good progress. I'll probably be AFK until Saturday. I can review more then.

Hi, sorry started replying before pushing the PR. Thanks for the review

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,57 @@ drop i cs0 = drop' i cs0
+--
+-- @since 0.11.2.0
+dropEnd :: Int64 -> ByteString -> ByteString
+dropEnd i p | i <= 0 = p
+dropEnd i p = go D.empty p
```
Style nit:

```suggestion
dropEnd i p | i <= 0 = p
dropEnd i p          = go D.empty p
```

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -696,6 +726,57 @@ drop i cs0 = drop' i cs0
+        -- for which we know they won't be dropped
+        getOutput :: ByteString -> D.Deque -> (ByteString, D.Deque)
+        getOutput out deque = case D.popFront deque of
+            Nothing                       -> (reverseChunks out, deque)
+            Just (x, deque') | D.elemLength deque' >= i ->
+                            getOutput (Chunk x out) deque'
+                             | otherwise  -> (reverseChunks out, deque)
```
```suggestion
            Just (x, deque') | D.elemLength deque' >= i ->
                            getOutput (Chunk x out) deque'
            _ -> (reverseChunks out, deque)
```

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -722,6 +803,23 @@ takeWhile f = takeWhile'
             n | n < S.length c -> Chunk (S.take n c) Empty
               | otherwise      -> Chunk c (takeWhile' cs)
 
+-- | Returns the longest (possibly empty) suffix of elements
+-- satisfying the predicate.
+--
+-- @'takeWhileEnd' p@ is equivalent to @'reverse' . 'takeWhile' p . 'reverse'@.
```
An example would be nice to have here.

3kyro replied:
I've used `Chunk (pack [1,2]) (Chunk (pack [3,4,6])) Empty` as an example of a lazy bytestring. Hope it's not too verbose, but textual representation of a bytestring is always tricky.

sjakobi replied:
I don't think that would be helpful for users who will mostly not be aware of `ByteString`'s internal constructors. If you want to represent the bytes as numbers, you can use the `OverloadedLists` syntax, e.g. `[1,2,3,4,6]`.

sjakobi commented on Data/ByteString/Lazy.hs:
```diff
@@ -733,6 +831,25 @@ dropWhile f = dropWhile'
             n | n < S.length c -> Chunk (S.drop n c) cs
               | otherwise      -> dropWhile' cs
 
+-- | Similar to 'P.dropWhileEnd',
+-- drops the longest (possibly empty) suffix of elements
+-- satisfying the predicate and returns the remainder.
+--
+-- @'dropWhileEnd' p@ is equivalent to @'reverse' . 'dropWhile' p . 'reverse'@.
```
An example would be nice to have here, too.

sjakobi commented on Data/ByteString/Lazy/Internal/Deque.hs:
```diff
@@ -0,0 +1,65 @@
+data Deque = Deque
+    { front :: [S.ByteString]
+    , rear :: [S.ByteString]
+    , -- | Accumulated length of deque's elements
+      elemLength :: !Int64
```
```suggestion
    , -- | Total length in bytes
      byteLength :: !Int64
```

sjakobi reviewed (commented):
A few more comments.

Since the implementation has turned out to be so complex, I'm relying on the tests to catch any bugs.

sjakobi reviewed (commented):
One more wibble – LGTM apart from that. :)

Bodigrim commented:
@3kyro please check CI failure.

3kyro commented:
> One more wibble – LGTM apart from that. :)

Hi @sjakobi, I've updated the examples using `OverloadedList`, however I still believe that we could have mentioned `Chunk` and `Empty`; for reference, "learn yourself haskell for real good" which is a pretty well known introduction to haskell explicitly mentions the lazy bytestrings constructors, eg in [input and output](http://learnyouahaskell.com/input-and-output).

Thanks again for the review :)

sjakobi reviewed (approved):
Good job!
